### PR TITLE
fix(git): set `--recurse-submodules` flag for checkout

### LINKED ISSUE #26057: Renovate does not update git submodules when changing branch
### Discussed in https://github.com/renovatebot/renovate/discussions/25689

<div type='discussions-op-text'>

<sup>Originally posted by **mtardy** November 10, 2023</sup>
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.54.3

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

I created a minimal reproduction repository here https://github.com/mtardy/renovate-submodules-repro.

---

I'm maintaining a project with Renovate that has a git submodule (and use this submodule in go.mod replace directive). Thanks to the option [cloneSubmodules](https://docs.renovatebot.com/configuration-options/#clonesubmodules) it works when I'm working on the main branch.

But when renovate needs to work on a stable branch, let's say v1.0, it `git checkout v1.0` but does not update the submodule along (with `git submodule update --init <module>` for example). So I end up on the `v1.0` stable branch with the content of the submodule folder corresponding to the one on the main branch. It then creates a PR with the changes and thus bump the submodule version to the one of the main branch because it never updated it to the correct commit.

I hope this a clear explanation! Is there a way to fix that behavior? Apparently, there are ways to immediately update submodules on checkout:
- https://stackoverflow.com/a/55631474/4561420 
- https://stackoverflow.com/a/49427199/4561420 
- tl;dr -> `git config submodule.recurse true`.

Is there a way to add git configs to the renovate run to pass `git config submodule.recurse true`?

I can do a minimal reproduction repository but I guess this explanation is clear enough maybe? Thanks!!

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I created a minimal reproduction repository here https://github.com/mtardy/renovate-submodules-repro.</div>

### PR DESCRIPTION
**Pull request by mtardy**
## Changes

Set `--recurse-submodules` when checking-out branches, so submodules are updated automatically when switching base branches.

## Context

- Closes https://github.com/renovatebot/renovate/issues/26057.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

File: lib/util/git/index.ts [hunk 1]
```diff
@@ -520,7 +520,9 @@ export async function checkoutBranch(
   logger.debug(`Setting current branch to ${branchName}`);
   await syncGit();
   try {
-    await gitRetry(() => git.checkout(['-f', branchName, '--']));
+    await gitRetry(() =>
+      git.checkout(['-f', '--recurse-submodules', branchName, '--']),
```
Comment (viceice) [inline lib/util/git/index.ts:524]: what happens if submodules are not initialized?

Comment (mtardy) [reply lib/util/git/index.ts:524]: If there is no submodules in the repo, this flag does nothing. Just try it in a submodule-free repo.

We could make this conditional on cloneSubmodules but I doubt it's necessary.

Comment (viceice) [reply lib/util/git/index.ts:524]: yes, make it conditional on `cloneSubmodules`

Comment (viceice) [reply lib/util/git/index.ts:524]: otherwise it will cause submodules clone

Comment (mtardy) [reply lib/util/git/index.ts:524]: mmmh indeed.

**Comment by mtardy** [comment]
Sorry I force pushed out of habit!

**Review by viceice — DISMISSED** [review]

**Comment by mtardy** [comment]
> this probably needs a new test case now

I don't get it, I've been trying to write a test case for hours now, I don't really understand what's happening:


```ts
      const repo = Git(base.path);
      console.log(await base.path)

      const submoduleBasePath = base.path + "/submodule"
      await fs.mkdir(submoduleBasePath)
      const submodule = Git(submoduleBasePath)
      await submodule.init()
      await submodule.addConfig('user.email', 'Jest@example.com');
      await submodule.addConfig('user.name', 'Jest');

      await fs.writeFile(submoduleBasePath + '/init_file', 'init');
      await submodule.add('init_file');
      await submodule.commit('init submodule');
      await submodule.branch(['initial'])

      await repo.submoduleAdd("./submodule", "./submodule")
      await repo.commit('add submodule');
      await repo.branch(["stable"])

      await fs.writeFile(submoduleBasePath + '/current_file', 'current');
      await submodule.add('current_file');
      await submodule.commit('update');
      await repo.add('submodule')
      await repo.commit('update submodule');

      expect((await repo.status()).isClean()).toBeTrue();
      await repo.checkout('stable')
      expect((await repo.status()).isClean()).toBeFalse();
      await repo.checkout('main')

      await repo.checkout(['--recurse-submodules', 'stable'])
      expect((await repo.status()).isClean()).toBeTrue();
```

This tests works fine. But as soon I'm using the `git` functions it doesn't work:

```ts
      await git.initRepo({
        cloneSubmodules: false,
        url: base.path,
      });
      await git.syncGit()
      expect((await git.getRepoStatus()).isClean()).toBeTrue();
      await git.checkoutBranch('stable');
      expect((await git.getRepoStatus()).isClean()).toBeFalse();
```

I don't really understand what is the `git.initRepo` really doing, when using checkout it does not really checkout the underlying repository.

**Review by viceice — DISMISSED** [review]

**Comment by mtardy** [comment]
> I think my new test has side effects on the others, can you help me @viceice ?

I think this should be enough https://github.com/renovatebot/renovate/pull/26163/commits/c3e34127fc3e5c534f6c959dd51d6938b6f713e3.

**Comment by mtardy** [comment]
I think my test is good enough now 🤠

**Review by viceice — APPROVED** [review]

**Pull request merged**